### ISSUE TITLE
should `tagName` be uppercase?

### ISSUE THREAD

**Issue by ianstormtaylor** [inikulin/parse5#210]
When working with on [Slate](https://github.com/ianstormtaylor/slate), we're trying to let people use `DOMParser` if it exists, or use `parse5` as a fallback they pass in if not (server-side, older browsers, etc.). But one of the big inconsistencies we're seeing is that native DOM parsing will give you tag names in uppercase, where `parse5` gives you lowercase.

Is this an issue with the native parsers? Wouldn't it be nice for `parse5` to return the same uppercase value for interoperability?

I'm probably missing something, so I figured I'd ask. Thanks!

**Comment by inikulin** [comment]
I guess you are using DOMParser in XML mode (which is default). If you switch it to HTML mode you will get consistent results with parse5.

**Comment by ianstormtaylor** [comment]
@inikulin ah wow, thank you! Very dumb mistake 😄

**Comment by ianstormtaylor** [comment]
@inikulin actually I misunderstood your answer. I think this issue is still applicable...

If you compare these two invocations:

```js
var parse5 = require("parse5")
var fragment = parse5.parseFragment('<div></div>')
console.log(fragment.childNodes[0].tagName)

// "div"
```

```js
var parser = new DOMParser()
var fragment = parser.parseFromString('<div></div>', 'text/html')
console.log(fragment.childNodes[0].childNodes[1].childNodes[0].tagName)

// "DIV"
```

- The `tagName` returned natively is uppercase, but with `parse5` it is lowercase.
- The structure of the fragment includes wrappers natively, but with `parse5` it does not (hence the need for the extra `.childNodes[0]` in the second example).

Am I still missing something?

**Comment by inikulin** [comment]
Oh, I see, sorry for misleading you. The problem is that parse5 produces AST and not full-featured DOM. HTML parser should lowercase all tag names with few exceptions for some foreign content elements. The tricky part is that `tagName` property is not the same thing as `tagName` property in `Element` interface.  When parser [creates element](https://html.spec.whatwg.org/multipage/parsing.html#creating-and-inserting-nodes) it passes `tagName` as `localName` argument to element [creation algorithm](https://dom.spec.whatwg.org/#concept-create-element) in DOM spec. Then in `Element` interface in DOM `tagName` [defined](https://dom.spec.whatwg.org/#dom-element-tagname) as uppercased version of `localName` in case of HTML elements. 

**tl;dr** `tagName` in parse5 AST is actually a `localName` in terms of DOM spec. However, since this is a DOM-agnostic implementation of parsing algorithm and to not confuse users who are not familiar with the difference (including me 10 minutes ago) it's called `tagName`.

**Comment by RReverser** [comment]
And to answer the second question:

> The structure of the fragment is wrapped by <html><body> natively, but with parse5 it does not (hence the need for the extra .childNodes[0] in the second example).

Reason is DOMParser parses provided HTML as a document not a fragment, which is just `.parse` in parse5.

**Comment by ianstormtaylor** [comment]
@inikulin @RReverser thank you! Using `.parse` instead of `.parseFragment` makes sense.

For the `tagName` issue though, does that mean `parse5` is not intended as a drop-in for `DOMParser`? I assumed from the docs that it was trying to be?

**Comment by RReverser** [comment]
Definitely not "drop-in" with defaults at least because the output formats are different (actual DOM nodes vs JS-objects-based-AST), but you can provide custom tree builders to parse5, including one that would create real DOM nodes or at least objects as similar to real DOM nodes as you need (uppercase tagnames etc.).

**Comment by ianstormtaylor** [comment]
Would `parse5` be open to implementing `tagName` as commonly known and putting the current `tagName` to the proper `localName` instead? I get the decision to use the "less confusing" name, but it seems like it's more confusing in the end if it doesn't match casing?

**Comment by inikulin** [comment]
@ianstormtaylor No, I have strong believe that we shouldn't introduce such change. In addition to already provided argument about confusion for users there is a technical reasons for that. This project implements HTML parsing specification which doesn't operate with term `localName` from DOM specification. parse5 uses tree adapters as an abstraction layer between parser and particular AST format. DOM is just one of such formats that parse5 can work with (see [jsdom](https://github.com/tmpvar/jsdom)) and transformation of `tagName` in `localName` is AST-specific and should be performed in particular tree adapter.

**Comment by inikulin** [comment]
@ianstormtaylor

> I get the decision to use the "less confusing" name, but it seems like it's more confusing in the end if it doesn't match casing?

In your case confusion comes from delusion that parse5 implements `DOMParser` functionality. When, in fact, it provides much more low level functionality. `DOMParser` requires at least full-featured DOM implementation working in conjunction with HTML parser. Therefore, current naming never caused confusion before, because users either use it for tasks that doesn't require full-featured DOM or use already mentioned jsdom if they need one.

**Comment by ianstormtaylor** [comment]
@inikulin very fair, thanks for explaining!
